instrumentation/aws-lambda-1.0/library/src/test/groovy/io/opentelemetry/instrumentation/awslambda/v1_0/AwsLambdaSqsMessageHandlerTest.groovy

```diff
@@ -0,0 +1,108 @@
+
+import static io.opentelemetry.trace.Span.Kind.CONSUMER
+import static io.opentelemetry.trace.Span.Kind.SERVER
+
+import com.amazonaws.services.lambda.runtime.Context
+import com.amazonaws.services.lambda.runtime.events.SQSEvent
+import io.opentelemetry.auto.test.InstrumentationSpecification
+import io.opentelemetry.auto.test.InstrumentationTestTrait
+import io.opentelemetry.trace.attributes.SemanticAttributes
+
+class AwsLambdaSqsMessageHandlerTest extends InstrumentationSpecification implements InstrumentationTestTrait {
+
+  private static final String AWS_TRACE_HEADER1 = "Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad8;Sampled=1"
+  private static final String AWS_TRACE_HEADER2 = "Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad9;Sampled=1"
+
+  class TestHandler extends TracingSQSMessageHandler {
+    @Override
+    protected void handleMessage(SQSEvent.SQSMessage event, Context context) {
+    }
+  }
+
+  def "messages with process spans"() {
+    when:
+    def context = Mock(Context)
+    context.getFunctionName() >> "my_function"
+    context.getAwsRequestId() >> "1-22-333"
+
+    def message1 = new SQSEvent.SQSMessage()
+    message1.setAttributes(["AWSTraceHeader": AWS_TRACE_HEADER1])
+    message1.setMessageId("message1")
+    message1.setEventSource("queue1")
+
+    def message2 = new SQSEvent.SQSMessage()
+    message2.setAttributes(["AWSTraceHeader": AWS_TRACE_HEADER2])
+    message2.setMessageId("message2")
+    message2.setEventSource("queue1")
+
+    def event = new SQSEvent()
+    event.setRecords([message1, message2])
+
+    new TestHandler().handleRequest(event, context)
+
+    then:
+    assertTraces(1) {
+      trace(0, 4) {
+        span(0) {
+          operationName("my_function")
+          spanKind SERVER
+          attributes {
+            "${SemanticAttributes.FAAS_EXECUTION}" "1-22-333"
+          }
+        }
+        span(1) {
+          operationName("queue1 receive")
+          spanKind CONSUMER
+          parentId(span(0).spanId)
+          attributes {
+            "${SemanticAttributes.MESSAGING_SYSTEM}" "AmazonSQS"
+            "${SemanticAttributes.MESSAGING_OPERATION}" "receive"
+          }
+          hasLink("5759e988bd862e3fe1be46a994272793", "53995c3f42cd8ad8")
+          hasLink("5759e988bd862e3fe1be46a994272793", "53995c3f42cd8ad9")
+        }
+        span(2) {
+          operationName("queue1 process")
+          spanKind CONSUMER
+          parentId(span(1).spanId)
+          attributes {
+            "${SemanticAttributes.MESSAGING_SYSTEM}" "AmazonSQS"
+            "${SemanticAttributes.MESSAGING_OPERATION}" "process"
+            "${SemanticAttributes.MESSAGING_MESSAGE_ID}" "message1"
+            "${SemanticAttributes.MESSAGING_DESTINATION}" "queue1"
+          }
+          hasLink("5759e988bd862e3fe1be46a994272793", "53995c3f42cd8ad8")
+        }
+        span(3) {
+          operationName("queue1 process")
+          spanKind CONSUMER
```

Comment:
So, we can have nested CONSUMER spans? But not CLIENT?

Reply:
Yeah our spec currently has an example of it

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md#batch-receiving

Since CONSUMER / PRODUCER aren't parent/child, I guess 1:1 doesn't matter as much. So in https://github.com/open-telemetry/opentelemetry-specification/issues/963 I'm also trying to clarify whether those kinds have any meaning other than INTERNAL anyways.

I'm happy to keep this instrumentation up to date as spec is updated, which I'm sure it will be :)